google-cloud-core/src/main/java/com/google/cloud/BaseServiceException.java

```diff
@@ -140,20 +141,36 @@ public BaseServiceException(IOException exception, boolean idempotent) {
     String location = null;
     String debugInfo = null;
     Boolean retryable = null;
-    if (exception instanceof GoogleJsonResponseException) {
-      GoogleJsonError jsonError = ((GoogleJsonResponseException) exception).getDetails();
-      if (jsonError != null) {
-        Error error = new Error(jsonError.getCode(), reason(jsonError));
-        code = error.code;
-        reason = error.reason;
-        retryable = isRetryable(idempotent, error);
-        if (reason != null) {
-          GoogleJsonError.ErrorInfo errorInfo = jsonError.getErrors().get(0);
-          location = errorInfo.getLocation();
-          debugInfo = (String) errorInfo.get("debugInfo");
+    if (exception instanceof HttpResponseException) {
+      if (exception instanceof GoogleJsonResponseException) {
+        GoogleJsonError jsonError = ((GoogleJsonResponseException) exception).getDetails();
+        if (jsonError != null) {
+          Error error = new Error(jsonError.getCode(), reason(jsonError));
+          code = error.code;
+          reason = error.reason;
+          retryable = isRetryable(idempotent, error);
+          if (reason != null) {
+            GoogleJsonError.ErrorInfo errorInfo = jsonError.getErrors().get(0);
+            location = errorInfo.getLocation();
+            debugInfo = (String) errorInfo.get("debugInfo");
+          }
+        } else {
+          code = ((GoogleJsonResponseException) exception).getStatusCode();
         }
       } else {
-        code = ((GoogleJsonResponseException) exception).getStatusCode();
+        // In cases where an exception is an instance of HttpResponseException but not
+        // an instance of GoogleJsonResponseException, check the status code to determine whether it's retryable
+        code = ((HttpResponseException) exception).getStatusCode();
+        Set<Error> retryableErrors = getRetryableErrors();
+        if (retryableErrors != null && !retryableErrors.isEmpty()) {
+          for (Error e : retryableErrors) {
+            if (e.getCode() != null && e.getCode().equals(code)) {
+              Error error = new Error(code, null);
```

Comment:
Yes, sorry, I missed the fact that "not assigning at all" is another kind of result. But the isRetriable() still has same type of the loop inside, so looping twice over same set of data looks suspicious. If you  can rewrite it without looping here again, it would be great.